src/lib_json/json_writer.cpp

```diff
@@ -309,31 +309,26 @@ static String valueToQuotedStringN(const char* value, unsigned length,
     // Should add a flag to allow this compatibility mode and prevent this
     // sequence from occurring.
     default: {
-      if (emitUTF8) {
-        result += *c;
-      } else {
-        unsigned int codepoint = utf8ToCodepoint(c, end);
-        const unsigned int FIRST_NON_CONTROL_CODEPOINT = 0x20;
-        const unsigned int LAST_NON_CONTROL_CODEPOINT = 0x7F;
-        const unsigned int FIRST_SURROGATE_PAIR_CODEPOINT = 0x10000;
-        // don't escape non-control characters
-        // (short escape sequence are applied above)
-        if (FIRST_NON_CONTROL_CODEPOINT <= codepoint &&
-            codepoint <= LAST_NON_CONTROL_CODEPOINT) {
-          result += static_cast<char>(codepoint);
-        } else if (codepoint <
-                   FIRST_SURROGATE_PAIR_CODEPOINT) { // codepoint is in Basic
-                                                     // Multilingual Plane
-          result += "\\u";
-          result += toHex16Bit(codepoint);
-        } else { // codepoint is not in Basic Multilingual Plane
-                 // convert to surrogate pair first
-          codepoint -= FIRST_SURROGATE_PAIR_CODEPOINT;
-          result += "\\u";
-          result += toHex16Bit((codepoint >> 10) + 0xD800);
-          result += "\\u";
-          result += toHex16Bit((codepoint & 0x3FF) + 0xDC00);
+      const auto appendHexChar = [&result](unsigned ch) {
+        result.append("\\u").append(toHex16Bit(ch));
+      };
+
+      unsigned codepoint = static_cast<unsigned>(*c);
+      if (codepoint > 0x7F && !emitUTF8) {
+        codepoint = utf8ToCodepoint(c, end);
+        if (codepoint < 0x10000) {
+          // codepoint is in Basic Multilingual Plane
+          appendHexChar(codepoint);
+        } else {
+          // codepoint is not in Basic Multilingual Plane
+          codepoint -= 0x10000;
+          appendHexChar(0xD800 + (codepoint >> 10));
+          appendHexChar(0xDC00 + (codepoint & 0x3FF));
         }
+      } else if (codepoint < 0x20) {
+        appendHexChar(codepoint);
+      } else {
+        result += static_cast<char>(codepoint);
```

Comment:
Mostly used range is `FIRST_NON_CONTROL_CODEPOINT ` to `FIRST_SURROGATE_PAIR_CODEPOINT `, it should be the first to judge.

Reply:
It is well efficient as it is. I know it might not be very obvious because tabs are rendered as only 2 spaces but I will give a quick explanation.


- First topmost `if` is taking care of any UTF-8 symbols(any character larger than 127) in case encoding is desired.
- In case encoding is not desired, the `else if` is triggered which will take care of encoding any non-printable characters and since an `unsigned` is used there is no need for lower bounds check.
- Lastly, any other symbol is directly added to the `result` which may include non-encoded UTF-8 characters is case `emitUTF8` is true.